apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1099,34 +1119,105 @@ private async Task ProcessPullMessagesAsync(List<ReceivedMessage> msgs, HashSet<
                 }
             }
 
-            private void HandleExtendLease(HashSet<string> msgIds)
+            private class LeaseCancellation : IDisposable
+            {
+                public LeaseCancellation(CancellationTokenSource softStopCts) =>
+                    _cts = CancellationTokenSource.CreateLinkedTokenSource(softStopCts.Token);
+
+                private readonly object _lock = new object();
+                private CancellationTokenSource _cts;
+
+                public CancellationToken Token
+                {
+                    get
+                    {
+                        lock (_lock)
+                        {
+                            return _cts?.Token ?? CancellationToken.None;
+                        }
+                    }
+                }
+
+                public void Dispose()
+                {
+                    lock (_lock)
+                    {
+                        _cts.Dispose();
+                        _cts = null;
+                    }
+                }
+
+                public bool IsDisposed
+                {
+                    get
+                    {
+                        lock (_lock)
+                        {
+                            return _cts == null;
+                        }
+                    }
+                }
+
+                public void Cancel()
+                {
+                    CancellationTokenSource cts2;
+                    lock (_lock)
+                    {
+                        cts2 = _cts;
+                    }
+                    cts2?.Cancel();
```

Comment:
Should this dispose as well? I'm a little lost as to what this is all doing.

Reply:
(Comments on this and Dispose would be useful.)

Reply:
Reading through everything else, it looks like it doesn't *have* to be disposed, but I suspect it would be good to do so.